File: sparkcql-core/src/main/scala/gov/va/sparkcql/adapter/data/DataAdapterAggregator.scala
package gov.va.sparkcql.adapter.data

import org.apache.spark.sql.{SparkSession, DataFrame}
import gov.va.sparkcql.logging.Log
import javax.xml.namespace.QName

sealed class DataAdapterAggregator(dataAdapters: List[DataAdapter]) extends DataAdapter {

  override val spark: SparkSession = null

  override def isDataTypeDefined(dataType: QName): Boolean = {
    dataAdapters.filter(_.isDataTypeDefined(dataType)).length > 0
  }
  
  override def acquire(dataType: QName): Option[DataFrame] = {
    val eligibleAdapters = dataAdapters.filter(_.isDataTypeDefined(dataType))
    if (!eligibleAdapters.isEmpty) {
      val acquiredDf = eligibleAdapters.flatMap(a => {
        val df = a.acquire(dataType)
        if (df.isEmpty) {
          Log.warn(s"${a.getClass().getSimpleName()} stated support for data type '${dataType.toString()}' but none was found.")
          None
        } else if (df.get.schema.fields.length == 0) {
          Log.warn(s"${a.getClass().getSimpleName()} returned a columnless dataframe when None should have been returned. Ignoring output.")
          None
        } else {
          df
        }
      })
      Some(acquiredDf.reduce(_.union(_)))   // union all dataframes
    } else {
      Log.warn(s"Attempted to acquire missing data for type '${dataType.toString()}' without verifying data was present.")
      None
    }
  }
}

File: sparkcql-core/src/main/java/gov/va/sparkcql/adapter/data/DataAdapterAggregator.java
package gov.va.sparkcql.adapter.data;

import java.util.List;

import javax.xml.namespace.QName;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;

public class DataAdapterAggregator implements DataAdapter {

    List<DataAdapter> dataAdapters;

    public DataAdapterAggregator(List<DataAdapter> dataAdapters) {
        this.dataAdapters = dataAdapters;
    }

    @Override
    public Boolean isDataTypeDefined(QName dataType) {
        return dataAdapters.stream().filter(a -> a.isDataTypeDefined(dataType)).count() > 0;
    }

    @Override
    public Dataset<Row> acquireData(QName dataType) {
        List<DataAdapter> eligibleAdapters = dataAdapters.stream().filter(a -> a.isDataTypeDefined(dataType)).toList();
        if (eligibleAdapters.size() > 0) {
            var acquiredDS = eligibleAdapters.stream().map(adapter -> {
                var ds = adapter.acquireData(dataType);
                if (ds == null) {
                    throw new RuntimeException(adapter.getClass().getSimpleName() + " stated support for data type '" + dataType.toString() + "' but none was found.");
                } else if (ds.schema().fields().length == 0) {
                    throw new RuntimeException(adapter.getClass().getSimpleName() + " returned a columnless dataframe when None should have been returned. Ignoring output.");
                } else {
                    return ds;
                }
            });
            var unionedDS = acquiredDS.reduce((a, b) -> a.union(b));
            return unionedDS.get();
        } else {
            throw new RuntimeException("Attempted to acquire missing data for type '" + dataType.toString() + "' without verifying data was present.");
        }
    }
    
}
